api-tests/tests/docker.rs

```diff
@@ -0,0 +1,91 @@
+mod common;
+mod container;
+
+use crate::common::{docker, Service};
+use crate::container::{
+    delete_app, deploy_app, logs, make_reqest, replicate_app, PREvant, Traefik,
+};
+use std::time::Duration;
+use testcontainers::Docker;
+
+#[tokio::test]
+async fn should_deploy_nginx() {
+    let docker = docker();
+    let traefik = docker.run(Traefik::default());
+    let prevant = docker.run(PREvant::default());
+
+    let app_name = deploy_app(
+        &prevant,
+        &vec![Service::new(
+            String::from("nginx"),
+            String::from("nginx:alpine"),
+        )],
+    )
+    .await
+    .expect("Should be able to deploy app");
+
+    let mut i = 0;
+    loop {
+        let response = make_reqest(&traefik, &app_name, "nginx").await;
+
+        if response.text().await.unwrap().contains("Welcome to nginx!") {
+            break;
+        }
+        std::thread::sleep(Duration::from_secs(5));
+
+        i += 1;
+    }
+
+    assert!(i < 5, "Cannot make request to nginx after {} attempts", i);
+
+    delete_app(&prevant, &app_name)
+        .await
+        .expect("Should be able to delete app");
+}
+
+#[tokio::test]
+async fn should_replicate_mariadb_with_replicated_env() {
+    let docker = docker();
+    let _traefik = docker.run(Traefik::default());
+    let prevant = docker.run(PREvant::default());
+
+    let db_service = Service::new(String::from("db"), String::from("mariadb:10.3.17"))
+        .with_replicated_env(
+            String::from("MYSQL_RANDOM_ROOT_PASSWORD"),
+            String::from("yes"),
+        );
+
+    let app_name = deploy_app(&prevant, &vec![db_service])
+        .await
+        .expect("Should be able to deploy app");
+
+    let replicated_app_name = replicate_app(&prevant, &app_name)
+        .await
+        .expect("Should be able to replicate app");
+
+    let mut i = 0;
+    loop {
+        if let Ok(logs) = logs(&prevant, &replicated_app_name, "db").await {
+            if logs.contains("GENERATED ROOT PASSWORD") {
+                break;
+            }
+        };
+
+        std::thread::sleep(Duration::from_secs(15));
```

Comment:
there is no workaround for thread sleep here?

Reply:
I wasn't able to get [`futures_retry`](https://docs.rs/futures-retry/0.5.0/futures_retry/) running and there are no good retry-until-APIs I know. Therefore, it was easier to work with sleep.